### DISCUSSION TITLE
Multiple PXE options in dnsmasq (proxyDHCP) causes iPXE self-loading loop (instead of loading boot script)

### DISCUSSION THREAD

**Discussion by henryptung** [ipxe/ipxe#909] [General]
I have a proxyDHCP setup using dnsmasq, where it provides both iPXE (via boot/snponly.efi) and grub2 as PXE options. I also have the iPXE script being loaded via a separate DHCP tag, as follows:

```
dhcp-match=set:iPXE,175

pxe-service=tag:!iPXE,x86PC,iPXE (legacy),boot/undionly.kpxe
pxe-service=tag:!iPXE,IA32_EFI,iPXE (EFI),boot/snponly.efi
pxe-service=tag:!iPXE,X86-64_EFI,iPXE (EFI),boot/snponly.efi
pxe-service=tag:!iPXE,BC_EFI,iPXE (EFI),boot/snponly.efi

pxe-service=tag:!iPXE,x86PC,GRUB 2 (i386-pc),boot/grub/i386-pc/core
pxe-service=tag:!iPXE,IA32_EFI,GRUB 2 (i386-efi),boot/grub/i386-efi/core.efi
pxe-service=tag:!iPXE,X86-64_EFI,GRUB 2 (x86_64-efi),boot/grub/x86_64-efi/core.efi
pxe-service=tag:!iPXE,BC_EFI,GRUB 2 (x86_64-efi),boot/grub/x86_64-efi/core.efi

dhcp-boot=tag:iPXE,boot/winpe.ipxe

log-dhcp
dhcp-no-override
dhcp-range=192.168.2.10,proxy
```

If I comment out the GRUB options and leave only a single PXE item, this works (i.e. iPXE loads, and then loads `boot/winpe.ipxe`. However, if I leave the GRUB options and choose iPXE from the PXE menu, iPXE instead detects `boot/snponly.efi` as its "boot script", and loops loading itself infinitely.

[captures.zip](https://github.com/ipxe/ipxe/files/10889771/captures.zip)

Note that in the attached packet captures, the DHCP options contain the correct script name in both cases - but for some reason, it is picked up in the one-option case, but ignored in the two-option case. As far as I can tell, the second (iPXE-driven) DHCP flow is identical in the two cases, so the behavior change seems to come from iPXE itself.

Note also that the two-option packet capture loops, because the loaded `snponly.efi` continues trying (and failing) to load an iPXE script, reloading itself instead. I stopped it after about 3 loops.

**Comment by NiKiZe** [comment]
Moving to discussion as I don't see this as an issue in iPXE, if it is we can create a detailed issue at that time.

Do you have any data in your main DHCP for filename?
What does th config menu in iPXE show?

Please be aware that the legacy menu system for PXE is rarely implemented for EFI. You might see different issues on different systems with that config.

**Reply by henryptung** [reply]
> Do you have any data in your main DHCP for filename?

No - the main DHCP sends "next server: 192.168.2.10" and "opt 66: 192.168.2.10", but otherwise sends no information about boot file name (neither GRUB files, snponly,efi, nor winpe.ipxe).

> What does th config menu in iPXE show?

I've copied the values for filename in the main menu (and various submenus) in the following scenarios; let me know if you want to see anything else.

(note: netX is present in all, but has no data)

single-option before loading script (Ctrl-B):
- filename: boot/snponly.efi
- net0/filename: \<not specified>
- net0/dhcp/filename: \<not specified>
- proxydhcp/filename: boot/snponly.efi

single-option after loading script (script just has `shell`):
- filename: boot/winpe.ipxe
- net0/filename: \<not specified>
- net0/dhcp/filename: \<not specified>
- proxydhcp/filename: boot/winpe.ipxe

double-option before reloading snponly.efi (Ctrl-B before cycling):
- filename: boot/snponly.efi
- net0/filename: \<not specified>
- net0/dhcp/filename: \<not specified>
- proxydhcp/filename: \<not specified>
- pxebs/filename: boot/snponly.efi

double-option after reloading snponly.efi (Ctrl-B after one cycle):
- filename: boot/snponly.efi
- net0/filename: \<not specified>
- net0/dhcp/filename: \<not specified>
- \<proxydhcp menu absent>
- pxebs/filename: boot/snponly.efi

Roughly speaking, it looks like the proxyDHCP result is being overridden by the follow-up DHCP in the first case, but in the second the pxebs result is "inherited" from the initial environment (to the point of ignoring the received filename altogether?).

FWIW, https://github.com/ipxe/ipxe/discussions/750 seems related - and there seems to be the same core point here. It's understandable that pxebs is something iPXE doesn't support in UEFI, and that it should ignore pxebs - however, if iPXE itself happens to _be loaded_ via pxebs, it seems to store that information in memory rather than ignore it (and then, ignore more relevant information from subsequent proxyDHCP requests).

I don't think iPXE should or needs to support pxebs (particularly in UEFI mode), but inheriting pxebs info from the boot environment, not overriding/replacing it on DHCP, and then reloading itself because of stale pxebs information seems like undesirable behavior. If iPXE isn't meant to support pxebs in UEFI, should it not load that config subtree at all?

**Reply by NiKiZe** [reply]
netX is a pointer to the netx that was last opened, since you have not opened any it is empty.

Not sure yet what is causing your issue. But will still mention, that if you have conflicting DHCP configs then you might get issues, this is not something that iPXE will fix. Need to check more exactly what is going on here.

With that said, in efi you might instead want to use autoexec.ipxe on your TFTP

**Reply by henryptung** [reply]
> that if you have conflicting DHCP configs then you might get issues

If you see anything to that effect in the captures or the info above, let me know - but, as far as I know, I'm using a proxyDHCP setup based on a (simplified) version of https://ipxe.org/appnote/proxydhcp that:

- doesn't check 175 suboptions for iPXE feature-set (assume that's not critical here)
- does use 175 to distinguish between iPXE and legacy PXE clients, as in example
- adds multiple pxe-service options, alongside iPXE

The only thing I have on my main DHCP is a next-server: 192.168.2.10 and tftp-server: 192.168.2.10, to handle some buggy PXE clients that would try to hit 192.168.2.1 instead. As far as I understand, that's a reasonable proxyDHCP setup and not a "conflicting DHCP config"; but if it is conflicting, happy to fix and test.

**Reply by stappersg** [reply]
From that URL `appnote/proxydhcp.txt · Last modified: 2021/08/05 20:53 by nikize`

> happy to fix and test.

Please do.

**Comment by stappersg** [comment]
> and loops loading itself infinitely.

https://ipxe.org/howto/chainloading#breaking_the_loop_with_the_dhcp_server

**Comment by mcb30** [comment]
So, based on a quick look at the packet captures, I **think** this is an interesting consequence of the fact that we cache and apply the DHCP packets from the previous boot stage.

@henryptung Your DHCP server is configured to provide the previous boot stage with a full PXEBS menu.  This will result in iPXE having a value for `${pxebs/filename}` when it starts up, before you even attempt DHCP.

Your DHCP server is then configured to provide iPXE with only a ProxyDHCPOFFER complete with filename.  This will cause iPXE to skip the ProxyDHCPACK and PXEBS stages as being unnecessary, since the filename is already present.

I suspect that you will therefore end up with the correct filename in `${proxydhcp/filename}`, but that this will be overridden by the previously cached filename in `${pxebs/filename}`.

The fix is probably going to be to have iPXE unregister any existing `pxebs` or `proxydhcp` settings blocks at the point that it completes a successful DHCPACK.  This potentially changes behaviour for some very obscure corner cases involving multiple network interfaces (e.g. for multipath SAN boot connections).

**Reply by henryptung** [reply]
....that makes sense, and that also suggests that I might be able to work around it by adding a dummy option after the iPXE script (to force it to return pxebs twice). Not convenient to test right now, but I can do it later tonight and report results.

**Reply by mcb30** [reply]
@henryptung I have reraised the issue as #913 and created pull request #914 as a candidate fix

**Reply by henryptung** [reply]
FWIW, I tried testing using PXE options in the boot-script loading phase, but they didn't seem to stick/work for iPXE script loading, so that was a dead end. Will report on the PR re: testing the fix.
